README.md
# <img align="left" height="50" src="https://raw.githubusercontent.com/strategitica/strategitica/main/dist/img/logo.png"> Strategitica

## A calendar view for [Habitica](https://habitica.com/), by [@misscurmudgeon](https://habitica.com/profile/88aa06bb-ec69-43d8-b58a-e6df4aa608ac)

> Welcome to Strategitica, a land conquered by a Habitican warrior during a time of great struggle in the Tasks War. After months of fighting a losing battle, the warrior realized that the only way to win was to somehow anticipate the enemy's movements. After much questing, the warrior found sacred tomes detailing how to harness the power of the legendary Habitican API. With this newfound power, the warrior commanded the API to foretell the enemy's movements. Using a fabled tactic known as a "calendar", the warrior organized this information so that each day's onslaught for the coming weeks could be examined in great detail. In no time at all, the warrior vanquished the evil tasks and colonized the territory the tasks abandoned. As the land was founded on the now revered principles of Information and Organization, the warrior dubbed this new land Strategitica.

## How to Use

Just go to https://strategitica.github.io/strategitica/ and enter your Habitica user ID and API token. Both can be found at https://habitica.com/user/settings/api. You may also go to https://strategitica.github.io/strategitica/?id=[YOUR-HABITICA-USER-ID] to pre-fill your user ID, if this saves time for you. After entering your info, you should see your tasks in calendar form for the next 90 days.

**Note:** Once your ID/token are entered, if you think Strategitica might not be up-to-date (for example, maybe you left Strategitica's page alone for a day, or you made changes on habitica.com or the Habitica app after using Strategitica, etc.), **you can refresh your data by clicking Refresh in the menu**. This saves you from having to refresh the page and re-enter your ID/token every time you want up-to-date info. The page does data refreshes itself immediately after you make changes using Strategitica, but it can't detect changes made outside of Strategitica.

## Optional Features

### Task Duration

Strategitica can show a task's duration to the right of the task. It can also show the total duration for all tasks in a day. To use this feature, create a tag for each task duration you want to use. Each tag should end with the following blurb:

> [strategitica|duration|##:##]

For example, if you add a tag named "[strategitica|duration|00:15]" (without quotes) to a task, Strategitica will recognize this as a 15-minute task. You can add whatever text you want before the blurb--it's only important that the blurb is at the very end. For example, "My tag name here [strategitica|duration|00:15]" will work just the same.

**Note:** As shown above, this only works for hh:mm format. For example, a tag for an hour should end with [strategitica|duration|01:00], not [strategitica|duration|1:00]. Something like [strategitica|duration|01:00:00] is also not supported.

**Another Note:** If a task has more than one duration tag, Strategitica won't recognize that the task has a duration **unless all of that task's duration tags are for the same duration**.

### Task Time of Day ###

Similarly to task durations, you can add a blurb to the end of a tag to tell Strategitica that a task should be completed in the morning, afternoon or evening. Tasks will be grouped by time of day.

> [strategitica|morning]

> [strategitica|afternoon]

> [strategitica|evening]

As with task durations, you can add whatever text you want before the blurb.

**Note:** morning, afternoon and evening are the only supported times of day.

**Another Note:** If a task has more than one time of day tag, Strategitica won't recognize that the task has a time of day. Use only one time of day tag on a task.

### "One-Time Dailies" ###

Habitica users will probably know that there are dailies (recurring tasks) and to-do's (one-time tasks). Both have due dates, but as far as I can tell, there is no penalty for not completing a to-do on time like there is with dailies. If you're like me and you need that penalty as motivation to get your one-time tasks done, then "one-time dailies" might be for you. As with task duration tags and time of day tags, create a tag that ends with:

> [strategitica|1td]

Add this to a daily that's really just a one-time task, and Strategitica will know to only show that task on its due date, not on any of its recurring due dates.

### Custom Days Limit ###

If you want to show a different number of days in the future than the default of 90 days, you can add days=[desired number of days] to the Strategitica URL. For example, https://strategitica.github.io/strategitica/?days=30 will show 30 days in the future. This was implemented to help users who have a lot of tasks, which can cause Strategitica to take too long to load. The maximum number of days allowed is 180.

## Notable Limitations/Issues ##

1. [Strategitica doesn't run cron.](https://github.com/strategitica/strategitica/issues/11) I may implement this one day, but right now, if you start your day and cron hasn't run yet, you'll see yesterday's tasks as today's tasks in your calendar. So, use habitica.com or the Habitica app to run cron.

2. [You can't edit task checkboxes yet.](https://github.com/strategitica/strategitica/issues/8) You can check and uncheck existing checkboxes, you just can't create new ones or edit/remove existing ones. Sorry, I just haven't gotten around to this yet.

3. [Strategitica wasn't made with non-midnight cron in mind.](https://github.com/strategitica/strategitica/issues/20) I'm still thinking of how to handle this.

4. Strategitica does its very best to put all your recurring tasks in the right place on your calendar. However, if a daily has a due date in the past, there's a chance it'll get placed at a date slightly off from when it should actually appear. This is because when a due date is in the past, it uses Habitica's API to find the next due date, and sometimes that date is wrong. I couldn't tell you why because I didn't make the API. For example, I set a daily's date on April 11th, 2020, and it's due every 12 months on that day of the month. In February 2021, I should see that task show up on April 11th, 2021. For some reason though, the Habitica API comes back with April 10th, 2021 as the earliest "next due" date. So basically, Strategitica sees April 10th, 2021 as the first date for this task, and it also sees that it should show up on April 11th, 2021 as well, so it adds both dates to the calendar. I'm not sure if there's much I can do about this.

5. If you have a daily set to repeat monthly on a day of the month that doesn't exist for some month (e.g. on the 30th, but there's no 30th in February), your task won't show up in that month, although Habitica will consider it due on the last day of that month. The one exception is if you set it to repeat on the 31st. Then, Strategitica will understand you want it to be due on the last day of the month. Sorry! Maybe I can do something about this someday, I'm not sure.

## Debugging ##

If you're having a problem with Strategitica and want to report it, please do this:

1. Go to https://strategitica.github.io/strategitica/?logs=true. This enables more verbose logging of what's going on with Strategitica than just errors.
2. Do whatever you were doing that caused an problem for you. You should be able to see the logs by clicking Logs in the menu, or in the login window if you're not able to get past entering your user ID and API token.
3. Describe your issue to me and include the full contents of the logs. You can do this by either [creating an issue on GitHub](https://github.com/strategitica/strategitica/issues/new) (this is the preferred option), or by sending me a message in Habitica. Here's [a link to my Habitica profile](https://habitica.com/profile/88aa06bb-ec69-43d8-b58a-e6df4aa608ac) if you choose to go that route.


CHANGELOG.md
[//]: # (Here's how to do semantic versioning:)
[//]: # (First Release                              1.0.0)
[//]: # (Backward compatible bug fix                1.0.1)
[//]: # (Backward compatible new feature            1.1.0)
[//]: # (Changes that break backward compatibility  2.0.0)

### [Strategitica](https://github.com/strategitica/strategitica) Changelog

## 1.4.1 (11 May, 2021)

* [#52](https://github.com/iymeko/strategitica/issues/52) resolved - Tasks with multiple duration tags that are all the same duration will now be able to show that duration. Previously, if a task had more than one duration tag (regardless of their duration), all duration tags would be ignored.

## 1.4.0 (10 May, 2021)

* [#48](https://github.com/iymeko/strategitica/issues/48) resolved - Task duration and time of day now appear like other task attributes. The tags associated with them will still show up with the rest of the task's tags.
* [#53](https://github.com/iymeko/strategitica/issues/53) resolved - Dailies with weekly recurrence that have no days of the week selected will no longer show in the calendar.

## 1.3.1 (6 May, 2021)

* [#44](https://github.com/iymeko/strategitica/issues/44) resolved - Some user/task attributes weren't being handled correctly in terms of data types, which caused Strategitica to hang for some users.


## 1.3.0 (27 April, 2021)

* [#36](https://github.com/iymeko/strategitica/issues/36) resolved - The default number of days that will be shown in the calendar has been reduced to 90 from 186, and the days URL parameter has been added. For example, add ?days=30 to the Strategitica URL to show 30 days in the future. The maximum days allowed is 180, so any number above that will cause Strategitica to show 180 days in the future.
* Added logging capabilities for debugging purposes. See https://github.com/iymeko/strategitica#debugging

## 1.2.1 (21 February, 2021)

* [#16](https://github.com/iymeko/strategitica/issues/16) resolved - mage class should now show correctly (instead of wizard)

This change is in [pull request #38](https://github.com/iymeko/strategitica/pull/38).

## 1.2.0 (21 February, 2021)

* [#29](https://github.com/iymeko/strategitica/issues/29) resolved - emoji shortcodes are now supported

This change is in [pull request #34](https://github.com/iymeko/strategitica/pull/34).

## 1.1.0 (15 February, 2021)

* [#19](https://github.com/iymeko/strategitica/issues/19) resolved - To-do's now show in the calendar
* [#20](https://github.com/iymeko/strategitica/issues/20) resolved - All dailies' recurring due dates now show in the calendar, with a few potential limitations:
  - If a daily has a due date in the past, there's a chance it'll get placed at a date slightly off from when it should actually appear. This is because when a due date is in the past, Strategitica will use Habitica's API to find the next due date, and sometimes, for reasons unknown, that date is wrong. For example, If a daily's start date is on April 11th, 2020, and it's due every 12 months on that day of the month. In February 2021, that task should show up on April 11th, 2021. For some reason though, the Habitica API comes back with April 10th, 2021 as the earliest "next due" date. So basically, Strategitica sees April 10th, 2021 as the first date for this task, and it also sees that it should show up on April 11th, 2021 as well, so it adds both dates to the calendar. Not sure if there's much I can do about this.
  - If a daily is set to repeat monthly on a day of the month that doesn't exist for some month (e.g. on the 30th, but there's no 30th in February), that daily won't show up in that month, although Habitica will consider it due on the last day of that month. The one exception is if it's set to repeat on the 31st. Then, Strategitica will understand it should be due on the last day of the month. Unsure if anything can be done about this.

Changes above are in [pull request #27](https://github.com/iymeko/strategitica/pull/27).

## 1.0.2 (11 February, 2021)

* [#18](https://github.com/iymeko/strategitica/issues/18) resolved - Instead of having the user's API token as a URL parameter, there is now a login modal
* Updated readme to reflect above changes
* Reorganized JS a lot by switching to Node modules
* Other readme updates
* Fixed an issue with the exp bar showing the wrong info

Changes above are in [pull request #26](https://github.com/iymeko/strategitica/pull/21). Also:

* Stopped tags with long names from going outside the task modal on small screens

## 1.0.1 (4 February, 2021)

* [#17](https://github.com/iymeko/strategitica/issues/17) resolved - Fixed refresh link

## 1.0.0 (2 February, 2021)

* Initial commit


webpack.config.js
const path = require('path');
const HtmlWebpackPlugin = require('html-webpack-plugin');
const { CleanWebpackPlugin } = require('clean-webpack-plugin');
const CopyPlugin = require("copy-webpack-plugin");

module.exports = {
    mode: 'development',
    entry: './src/index.js',
    devtool: 'inline-source-map',
    plugins: [
        new CleanWebpackPlugin({ cleanStaleWebpackAssets: false }),
        new HtmlWebpackPlugin({
            title: 'Development',
        }),
        new CopyPlugin({
            patterns: [
                { from: 'src/img', to: 'img' }
            ],
        }),
    ],
    output: {
        filename: '[name].bundle.js',
        path: path.resolve(__dirname, 'dist'),
    },
    module: {
        rules: [
            {
                test: /\.css$/i,
                use: ['style-loader', 'css-loader'],
            },
            {
                test: /\.scss$/,
                use: [
                    { loader: 'style-loader' },
                    {
                        loader: 'css-loader',
                        options: {
                            sourceMap: true,
                            modules: {
                                auto: true,
                                localIdentName: '[local]_[hash:base64:5]'
                            }
                        }
                    },
                    {
                        loader: 'postcss-loader',
                        options: {
                            postcssOptions: {
                                config: 'postcss.config.js',
                            },
                            sourceMap: true,
                        },
                    },
                    {
                        loader: 'sass-loader', options: {
                            sassOptions: {
                                outputStyle: 'compressed'
                            },
                            sourceMap: true
                        }
                    }
                ]
            },
            {
                test: /\.(png|svg|jpe?g|jpeg|gif)$/i,
                type: 'asset/resource',
            },
            {
                test: /\.(woff|woff2|eot|ttf|otf)$/i,
                type: 'asset/resource',
            },
        ],
    },
};

src/modules/user.js
import * as Utils from './utils.js';
import $ from 'jquery';

export class User {
    /**
     * 
     * @param {string} id - The user ID
     * @param {string} token - The user's API token
     */
    constructor(id, token) {
        this.id = id;
        this.token = token;
        this.tags = {};
        this.tasks = {};
        this.name = '';
        this.displayName = '';
        this.class = '';
        this.level = 0;
        this.hp = 0;
        this.hpMax = 0;
        this.exp = 0;
        this.expToNextLevel = 0;
        this.mp = 0;
        this.mpMax = 0;
        this.isSleeping = false;
        this.dayStart = 0;
    }

    /**
     * @see {@link https://habitica.com/apidoc/#api-Tag-GetTag|Tag - Get a tag}
     * @see {@link https://habitica.com/apidoc/#api-Task-GetUserTasks|Task - Get a user's tasks}
     * @see {@link https://habitica.com/apidoc/#api-User-UserGet|User - Get the authenticated user's profile}
     */
    create(onComplete) {
        var user = this;

        try {
            $.when(this.getTags(), this.getDailies(), this.getTodos(), this.getUserInfo()).then(function (tags, dailies, todos, userInfo) {
                Utils.updateLogs('Data loaded successfully!');

                let userTagNames = {};

                if (tags[0].data !== null) {
                    tags[0].data.forEach(function (value) {
                        userTagNames[value.id.toString()] = value.name;
                    });
                }

                user.tags = userTagNames;

                var allTasks = {};

                [].concat(dailies[0].data, todos[0].data).forEach(function (item, index, myArray) {
                    allTasks[item.id] = item;
                });

                user.tasks = allTasks;
                
                if ('data' in userInfo[0]) {
                    user.setUserInfo(userInfo[0].data);
                }

                if (onComplete) {
                    onComplete();
                }
            }, function (error) {
                $('#strategitica-login-progress').addClass('d-none');

                var message = 'Couldn\'t get data';
                if ('status' in error) {
                    message += ': ' + error.status + ' Error';
                    if ('responseJSON' in error) {
                        if ('message' in error.responseJSON) {
                            message += ' - ' + error.responseJSON.message;
                        }
                    }
                }

                Utils.updateLogs(message, true);
            });
        }
        catch (error) {
            Utils.updateLogs('Couldn\'t get data: ' + error.responseText, true);
        }
    }

    getTags(onComplete) {
        var user = this;
        var options = {
            async: true,
            url: 'https://habitica.com/api/v3/tags',
            type: 'GET',
            dataType: 'json',
            contentType: 'application/json',
            cache: false,
            beforeSend: function (xhr) {
                xhr.setRequestHeader('x-client', Utils.appClient);
                xhr.setRequestHeader('x-api-user', user.id);
                xhr.setRequestHeader('x-api-key', user.token);
            }
        };

        if (onComplete) {
            try {
                $.ajax(options)
                .done(function (data) {
                    let userTagNames = {};

                    if (data.data !== null) {
                        data.data.forEach(function (value) {
                            userTagNames[value.id.toString()] = value.name;
                        });
                    }
    
                    user.tags = userTagNames;

                    Utils.updateLogs('Tags loaded successfully');

                    onComplete();
                })
                .fail(function (jqXHR, textStatus, errorThrown) {
                    var message = 'Couldn\'t get user\'s tasks';
                    if ('status' in jqXHR) {
                        message += ': ' + error.status + ' Error';
                        if ('responseJSON' in error) {
                            if ('message' in error.responseJSON) {
                                message += ' - ' + error.responseJSON.message;
                            }
                        }
                    }
    
                    Utils.updateLogs(message, true);
                })
            }
            catch (error) {
                Utils.updateLogs('Couldn\'t get user\'s tasks: ' + error.responseText, true);
            }
        }
        else {
            return $.ajax(options);
        }
    }

    getTasks(onComplete) {
        var user = this;

        try {
            $.when(this.getDailies(), this.getTodos()).then(function (dailies, todos) {
                var allTasks = {};
    
                [].concat(dailies[0].data, todos[0].data).forEach(function (item, index, myArray) {
                    allTasks[item.id] = item;
                });
    
                user.tasks = allTasks;

                Utils.updateLogs('Tasks loaded successfully');

                if(onComplete) {
                    onComplete();
                }
            }, function (error) {
                var message = 'Couldn\'t get tasks';
                if ('status' in error) {
                    message += ': ' + error.status + ' Error';
                    if ('responseJSON' in error) {
                        if ('message' in error.responseJSON) {
                            message += ' - ' + error.responseJSON.message;
                        }
                    }
                }
    
                Utils.updateLogs(message, true);
            });
        }
        catch (error) {
            Utils.updateLogs('Couldn\'t get tasks: ' + error.responseText, true);
        }
    }

    getDailies() {
        var userId = this.id;
        var apiToken = this.token;
        
        return $.ajax({
            async: true,
            url: 'https://habitica.com/api/v3/tasks/user',
            type: 'GET',
            dataType: 'json',
            contentType: 'application/json',
            cache: false,
            data: {
                type: 'dailys'
            },
            beforeSend: function (xhr) {
                xhr.setRequestHeader('x-client', Utils.appClient);
                xhr.setRequestHeader('x-api-user', userId);
                xhr.setRequestHeader('x-api-key', apiToken);
            }
        });
    }

    getTodos() {
        var userId = this.id;
        var apiToken = this.token;

        return $.ajax({
            async: true,
            url: 'https://habitica.com/api/v3/tasks/user',
            type: 'GET',
            dataType: 'json',
            contentType: 'application/json',
            cache: false,
            data: {
                type: 'todos'
            },
            beforeSend: function (xhr) {
                xhr.setRequestHeader('x-client', Utils.appClient);
                xhr.setRequestHeader('x-api-user', userId);
                xhr.setRequestHeader('x-api-key', apiToken);
            }
        });
    }

    getUserInfo(onComplete) {
        var user = this;
        var options = {
            async: true,
            url: 'https://habitica.com/api/v3/user',
            type: 'GET',
            dataType: 'json',
            contentType: 'application/json',
            cache: false,
            beforeSend: function (xhr) {
                xhr.setRequestHeader('x-client', Utils.appClient);
                xhr.setRequestHeader('x-api-user', user.id);
                xhr.setRequestHeader('x-api-key', user.token);
            }
        };

        if (onComplete) {
            try {
                $.ajax(options)
                .done(function (data) {
                    if ('data' in data) {
                        user.setUserInfo(data.data);
                    }

                    onComplete();
                })
                .fail(function (jqXHR, textStatus, errorThrown) {
                    var message = 'Couldn\'t get user info';
                    if ('status' in jqXHR) {
                        message += ': ' + error.status + ' Error';
                        if ('responseJSON' in error) {
                            if ('message' in error.responseJSON) {
                                message += ' - ' + error.responseJSON.message;
                            }
                        }
                    }
    
                    Utils.updateLogs(message, true);
                })
            }
            catch (error) {
                Utils.updateLogs('Couldn\'t get user info: ' + error.responseText, true);
            }
        }
        else {
            return $.ajax(options);
        }
    }

    setUserInfo(data) {
        var user = this;

        if (data !== null) {
            if ('auth' in data) {
                if ('local' in data.auth) {
                    if ('username' in data.auth.local) {
                        user.name = data.auth.local.username;
                    }
                }
            }

            if ('profile' in data) {
                if ('name' in data.profile) {
                    user.displayName = data.profile.name;
                }
            }
            
            if ('stats' in data) {
                if ('class' in data.stats) {
                    user.class = data.stats.class.toLowerCase() === 'wizard' ? 'Mage' : data.stats.class.charAt(0).toUpperCase() + data.stats.class.slice(1);
                }
                if ('lvl' in data.stats) {
                    user.level = data.stats.lvl;
                }
                if ('hp' in data.stats) {
                    user.hp = data.stats.hp;
                }
                if ('maxHealth' in data.stats) {
                    user.hpMax = Math.floor(data.stats.maxHealth);
                }
                if ('exp' in data.stats) {
                    user.exp = Math.floor(data.stats.exp);
                }
                if ('toNextLevel' in data.stats) {
                    user.expToNextLevel = data.stats.toNextLevel;
                }
                if ('mp' in data.stats) {
                    user.mp = Math.floor(data.stats.mp);
                }
                if ('maxMP' in data.stats) {
                    user.mpMax = data.stats.maxMP;
                }
            }

            if ('preferences' in data) {
                if ('sleep' in data.preferences) {
                    user.isSleeping = data.preferences.sleep;
                }
                if ('dayStart' in data.preferences) {
                    user.dayStart = data.preferences.dayStart;
                }
            }

            var message = `User info loaded successfully. Here's some info about you:<br>
                 Username: @${user.name}<br>
                 Display Name: ${user.displayName}<br>
                 Class: ${user.class}<br>
                 Level: ${user.level}<br>
                 HP: ${user.hp} / ${user.hpMax}<br>
                 Experience: ${user.exp} / ${user.expToNextLevel}<br>
                 MP: ${user.mp} / ${user.mpMax}<br>
                 Resting in the tavern: ${user.isSleeping}<br>
                 Day start: ${user.dayStart}<br>
                 Tags: ${Object.values(user.tags)}`;
            Utils.updateLogs(message);
        }
    }

    /**
     * Changes the user's tavern status to the opposite of whatever it
     * currently is.
     * 
     * @see {@link https://habitica.com/apidoc/#api-User-UserSleep|User - Make the user start / stop sleeping (resting in the Inn)}
     */
    changeTavernStatus(onComplete) {
        var user = this;

        try {
            $.ajax({
                async: true,
                url: 'https://habitica.com/api/v3/user/sleep',
                type: 'POST',
                data: !user.isSleeping,
                dataType: 'json',
                contentType: 'application/json',
                cache: false,
                beforeSend: function (xhr) {
                    xhr.setRequestHeader('x-client', Utils.appClient);
                    xhr.setRequestHeader('x-api-user', user.id);
                    xhr.setRequestHeader('x-api-key', user.token);
                }
            })
                .done(function (data) {
                    user.isSleeping = data.data;

                    var message = 'You have successfully ' + (user.isSleeping === true ? 'entered' : 'left') + ' the tavern.';
                    Utils.updateLogs(message);
                    Utils.updateToast('success', 'Tavern Status', message);

                    if (onComplete) {
                        onComplete();
                    }
                })
                .fail(function (jqXHR, textStatus, errorThrown) {
                    let message = 'Couldn\'t update tavern status: <br>' + jqXHR.status + ' Error';

                    if ('responseJSON' in jqXHR) {
                        if ('message' in jqXHR.responseJSON) {
                            message += ' - ' + jqXHR.responseJSON.message;
                        }
                    }

                    Utils.updateLogs('Error: ' + message, true);
                    Utils.updateToast('error', 'Error', message);
                });
        }
        catch (error) {
            $('#strategitica-tavern-change1').hide();

            var message = 'Couldn\'t update tavern status: <br>' + error.responseText;
            Utils.updateLogs('Error: ' + message, true);
            Utils.updateToast('error', 'Error', message);
        }
    }
}

src/modules/utils.js
import $ from 'jquery';

export const appClient = '88aa06bb-ec69-43d8-b58a-e6df4aa608ac-Strategitica';

export const monthNames = [
    ['January', 'Jan'],
    ['February', 'Feb'],
    ['March', 'Mar'],
    ['April', 'Apr'],
    ['May', 'May'],
    ['June', 'Jun'],
    ['July', 'Jul'],
    ['August', 'Aug'],
    ['September', 'Sep'],
    ['October', 'Oct'],
    ['November', 'Nov'],
    ['December', 'Dec']
];

export const frequencyPlurals = {
    'daily': 'day(s)',
    'weekly': 'week(s)',
    'monthly': 'month(s)',
    'yearly': 'year(s)'
};

// debulked onresize handler
export function onResize(c,t){onresize=function(){clearTimeout(t);t=setTimeout(c,100)};return c};

/**
 * Gets a URL parameter by its name.
 * @param {string} name - The parameter name
 */
export function getUrlParameter(name) {
    name = name.replace(/[\[]/, '\\[').replace(/[\]]/, '\\]');
    var regex = new RegExp('[\\?&]' + name + '=([^&#]*)');
    var results = regex.exec(location.search);
    return results === null ? '' : decodeURIComponent(results[1].replace(/\+/g, ' '));
}

export const showLogs = getUrlParameter('logs') === 'true';

let calendarDaysLimitFromParam = parseInt(getUrlParameter('days'));
let calendarDaysLimitMax = 180;
let calendarDaysLimitTemp = 90; // [1]

if (!isNaN(calendarDaysLimitFromParam)) {
    if (calendarDaysLimitFromParam > calendarDaysLimitMax) {
        updateLogs('Custom days limit (' + calendarDaysLimitFromParam + ') exceeds the maximum allowed (' + calendarDaysLimitMax + '); the default days limit will be used (' + calendarDaysLimitTemp + ')');
    }
    else {
        if (calendarDaysLimitFromParam < 0) {
            updateLogs('Custom days limit (' + calendarDaysLimitFromParam + ') cannot be less than zero; the default days limit will be used (' + calendarDaysLimitTemp + ')');
        }
        else {
            updateLogs('Custom days limit recognized (' + calendarDaysLimitFromParam + ')');
            calendarDaysLimitTemp = calendarDaysLimitFromParam; // [1]
        }
    }
}

export const calendarDaysLimit = calendarDaysLimitTemp;

export function escapeQuotes(string) {
    return string.replace(/'/g, '&apos;').replace(/"/g, '&quot;');
}

export function escapeHtmlBrackets(string) {
    return string.replaceAll('<', '&lt;').replaceAll('>', '&gt;');
}

/**
 * You can't just output a Date variable in JS and expect it to look readable,
 * and for debugging purposes, I want things to be readable. So this is here to
 * create understandable keys for the big fat date/task object in
 * {@link loadCalendar}. Also, I needed dates to be formatted in a way that an
 * input element with type="date" could work with.
 * @param {Date} date 
 * @returns {string} A readable version of a given date
 */
export function getDateKey(date) {
    return date.getUTCFullYear() + '-' + (date.getUTCMonth() <= 8 ? '0' : '') + (date.getUTCMonth() + 1) + '-' + (date.getUTCDate() <= 9 ? '0' : '') + date.getUTCDate();
}

/**
 * Gets the ordinal of a given number, e.g. the ordinal for 1 is "st", the
 * ordinal for 12 is "th", etc.
 * @param {number} number 
 * @returns {string} The ordinal
 */
export function getNumberOrdinal(number) {
    var numberAsString = number.toString();
    var ordinal = '';

    if (numberAsString.endsWith('11') || numberAsString.endsWith('12') || numberAsString.endsWith('13')) {
        ordinal = 'th';
    }
    else if (numberAsString.endsWith('1')) {
        ordinal = 'st';
    }
    else if (numberAsString.endsWith('2')) {
        ordinal = 'nd';
    }
    else if (numberAsString.endsWith('3')) {
        ordinal = 'rd';
    }
    else {
        ordinal = 'th';
    }

    return ordinal;
}

/**
 * Ahem, uh, in case you didn't know, some months have 30 days, some have 31...
 * some even have 28 or 29! This determines what the last day of the month is
 * for a given month and year.
 * @param {number} year - The year, i.e. 2020
 * @param {number} month - The month, e.g. 0 for January, 1 for February, etc.
 * @returns {number} The last day of the given month and year, e.g. 31
 */
export function getLastDayOfMonth(year, month) {
    return new Date(year, month + 1, 0).getDate();
}

export function formatDuration(durationInMinutes) {
    var durationFormatted = '';

    if (durationInMinutes >= 60) {
        var durationInHours = durationInMinutes / 60;

        durationFormatted = (durationInHours % 1 === 0 ? durationInHours : durationInHours.toFixed(1)) + 'h';
    }
    else {
        durationFormatted = durationInMinutes + 'm';
    }

    return durationFormatted;
}

export function updateToast(type, title, body, toastToHide) {
    var el = $('#strategitica-toast-' + type);
    el.find('.toast-title-js').html(escapeHtmlBrackets(escapeQuotes(title)));
    el.find('.toast-body-js').html(escapeHtmlBrackets(escapeQuotes(body)));

    el.toast('show');

    if (toastToHide) {
        var el2 = $('#strategitica-toast-' + toastToHide);
        el2.toast('hide');
    }
}

export function updateLogs(text, isError) {
    if (showLogs || isError) {
        var contextClasses = '';

        if (isError) {
            contextClasses += 'bg-danger text-light';
        }
        else  {
            contextClasses += 'bg-dark text-light'
        }
    
        var now = new Date();
        var timestamp = $('<span class="pr-1">[' + (now.getHours() < 10 ? '0' : '') + now.getHours() + ':' + (now.getMinutes() < 10 ? '0' : '') + now.getMinutes() + ':' + (now.getSeconds() < 10 ? '0' : '') + now.getSeconds() + ']</span> ');
        
        var textContainer = $('<code class="media px-1 ' + contextClasses + '" />')
        var mediaBody = $('<div class="media-body" />');
    
        mediaBody.append(text);
        textContainer.append(timestamp);
        textContainer.append(mediaBody);
        $('.strategitica-logs-js').append(textContainer);
    }
}